Supporting Ban Post or Comment outside Taiwanese IP

PichuChen opened the pull request:
Requirement:
```
- 新增看板權限限制非台灣IP發/推文 (實作難度: 易)
  Please use the country value returned from `fromd`.
  Please guard this via a compile time flag.
```

Changes:
1. Introduce a new compiler flag `POST_COMMENT_WITH_TAIWAN_IP_ONLY` for this function.
2. Implement `is_taiwan_ip()` which determine if current user is from Taiwan IP.
3. Insert checking Taiwan IP into `const char * postperm_msg(const char *bname)`.
4. Introduce a board configuration `BRD_TAIWAN_IP_ONLY (0x2000000)` for this function.

Screenshots:
<img width="1002" alt="截圖 2022-03-08 下午8 04 53" src="https://user-images.githubusercontent.com/600238/157236961-ef73dd4a-3cc1-4c06-a92d-1467831ba7f3.png">
<img width="1014" alt="截圖 2022-03-08 下午8 05 12" src="https://user-images.githubusercontent.com/600238/157236982-e0e0afb1-beed-4b1d-899e-fc7b316caa26.png">
<img width="1037" alt="截圖 2022-03-08 下午8 05 50" src="https://user-images.githubusercontent.com/600238/157236993-f2a6659b-4993-4ba1-b1ff-7b5d44dd0072.png">

Remark:
* Please enable `FROMD` and flag `POST_COMMENT_WITH_TAIWAN_IP_ONLY` when use this function.

TODO:
* Changing `from_cc` to ISO code instead of Big-5 string.
* Review the post or comment checking flow and write the document.

wens committed:
local: util/bbsrf: Check load before executing mbbsd

It makes sense to check the CPU load in bbsrf, the lightweight shim for
SSH connections, before executing the heavy mbbsd, which might end up
exiting right away due to high CPU load.

This alleviates CPU load during high bursts of connections.

wens committed:
fromd: Allow both GeoIP2 and GeoIP database usage

When GeoIP2 support was introduced, the Makefile was mistakenly written
to enable one or the other, when in fact the code supports enabling both
with GeoIP2 as the preferred database.

Fix the Makefile so that both solutions are enabled.

PichuChen committed:
support ban outside Taiwan IP from FROMD

kcwu commented on mbbsd/board.c:
```diff
@@ -496,6 +496,15 @@ b_config(void)
 		ANSI_COLOR(1) "�T�� " : "���\\ " );
 
+#ifdef POST_COMMENT_WITH_TAIWAN_IP_ONLY
+	// use '8' instead of '1', to prevent 'l'/'1' confusion
```
incorrect comment

PichuChen replied:
fixed, thanks

kcwu commented:
please squash your 4 commits into single one

robertabcd commented on include/proto.h:
```diff
@@ -101,6 +101,9 @@ void log_crosspost_in_allpost(const char *brd, const fileheader_t *postfile);
 int check_cooldown(boardheader_t *bp);
 #endif
+#ifdef POST_COMMENT_WITH_TAIWAN_IP_ONLY
+int is_taiwan_ip();
```
Please return `bool`.

robertabcd commented on mbbsd/bbs.c:
```diff
@@ -433,6 +433,26 @@ int IsFreeBoardName(const char *brdname)
+ */
+int is_taiwan_ip()
+{
+#ifdef POST_COMMENT_WITH_TAIWAN_IP_ONLY
```
Might just wrap this `#ifdef` around the whole function, so you don't need the `#else` part.

PichuChen replied:
Let me think twice. If I warp whole function, I will need to change the function name or the future developer might use this function without `POST_COMMENT_WITH_TAIWAN_IP_ONLY` flag and got compiler error unexpected.

robertabcd replied:
That should be a compiler error, right?

I don't think it's correct that `is_taiwan_ip()` always return `true` without `POST_COMMENT_WITH_TAIWAN_IP_ONLY`.

You also have a ifdef in the header file already.

PichuChen replied:
Yes, you're right.

kcwu commented on mbbsd/bbs.c:
```diff
@@ -433,6 +433,24 @@ int IsFreeBoardName(const char *brdname)
+ * Check if user ip address is from Taiwan
+ *
+ * This function will compare from_cc string and "Taiwan" in Chinese (Big5)
+ * from_cc will generate when user login bbs system, and set from FORMD
```
s/FORMD/FROMD/ ?

kcwu replied:
BTW, you have the same typo in the description of this pull request as well.

PichuChen replied:
Thank you 😣

kcwu commented on mbbsd/bbs.c:
```diff
@@ -433,6 +433,24 @@ int IsFreeBoardName(const char *brdname)
+ * Return true when user is from Taiwan IP, false when outside Taiwan
+ */
+
+#ifdef POST_COMMENT_WITH_TAIWAN_IP_ONLY
```
IMO, the function comment should be inside the #ifdef block.

PichuChen replied:
Sure.

wens committed:
mbbsd: Fix incorrect wording for "punishment"

kcwu commented on mbbsd/bbs.c:
```diff
@@ -433,6 +433,24 @@ int IsFreeBoardName(const char *brdname)
+is_taiwan_ip()
+{
+    // TODO: move this string to string table, eg: common.h
+    return !strncmp("�O�W", from_cc, STRLEN);
```
Given from_cc is properly null terminated, use strcmp instead of strncmp.
If you don't trust the null character and want to be extremely safe, use sizeof(from_cc) instead of STRLEN, just in case the size may be changed in the future.

PichuChen replied:
Agree, using `sizeof(from_cc)` is better.

wens committed:
mbbsd: register_sms: Fix misleading default for user agreement

The user agreement requires an explicit "yes" (agree) from the user.
However the prompt is misleading in that 'Y" is capitalized, implying
that it is the default. This leads to users just pressing enter, and
returning to the menu.

Change 'Y" to lower case to signal that an explicit 'y' is required.

wens committed:
util: Include 'var.h' for external symbols instead of declaring locally

Some global variables were declared locally without an extern qualifier.
This leads to link errors with newer toolchains.

Instead, drop the declarations and include "var.h" which provides all
the proper declarations.

IepIweidieng commented:
*Disclaimer: I'm not directly associated with any of the main maintainers of this repository.*

I'm not sure whether directly put `taiwan` in the name of functions, macros, etc. is elegant for implementing region restriction.
Displaying "Taiwan"/"臺灣" as the main region to the users of Ptt seems totally fine, however.

I think that it might be better if a work is done to add an option in `config.h` or `pttbbs.conf` which controls the region exempted from the restriction.
Assume the option will be named `STR_FROM_REGION_LOCAL` (or `ISO_CODE_REGION_LOCAL` and the likes). As an example, a server maintainer willing to change the exempted region from Taiwan to America can write `#define STR_FROM_REGION_LOCAL "美國"` (default: `"臺灣"`) in `pttbbs.conf` .

Although this is not an issue until some person or organization is willing to establish a BBS targeting people oversea using the PttBBS server code, so perhaps such work deserves another PR.

PichuChen commented:
> _Disclaimer: I'm not directly associated with any of the main maintainers of this repository._
> 
> I'm not sure whether directly put `taiwan` in the name of functions, macros, etc. is elegant for implementing region restriction. Displaying "Taiwan"/"臺灣" as the main region to the users of Ptt seems totally fine, however.
> 
> I think that it might be better if a work is done to add an option in `config.h` or `pttbbs.conf` which controls the region exempted from the restriction. Assume the option will be named `STR_FROM_REGION_LOCAL` (or `ISO_CODE_REGION_LOCAL` and the likes). As an example, a server maintainer willing to change the exempted region from Taiwan to America can write `#define STR_FROM_REGION_LOCAL "美國"` (default: `"臺灣"`) in `pttbbs.conf` .
> 
> Although this is not an issue until some person or organization is willing to establish a BBS targeting people oversea using the PttBBS server code, so perhaps such work deserves another PR.

I am just follow the requirement, but in technically, I will design for an IP address white list or black list, or an ASN white list or black list.

Btw, the requirement has been changed, they want to check (or ban) the user according there AOTP register status, so you can see this PR (https://github.com/ptt/pttbbs/pull/108)

hungte commented:
First of all, it is better to implement this as region-specific (like @IepIweidieng mentioned) feature instead of hard-coding "Taiwan".

Second, given the author said "the requirement has been changed", I'm going to close the PR.
